source/grid/grid_tools_cache.cc

```diff
@@ -83,6 +83,20 @@ namespace GridTools
 
 
 
+  template<int dim, int spacedim>
+  const std::map<unsigned int, Point<spacedim> > &
+  Cache<dim,spacedim>::get_used_vertices() const
+  {
+    if (update_flags & update_used_vertices)
+      {
+        used_vertices = GridTools::extract_used_vertices(*tria, *mapping);
+        update_flags = update_flags & ~update_used_vertices;
```

Comment:
i think that logic is a bit confusing, why not call `GridTools::extract_used_vertices` somewhere else once (if flag is set) and only return here? Then you don't need `mutable` and nor do you ned those tricks with `~update_used_vertices` .

Reply:
because the flags are invalidated by signals from the triangulation, and we don't want to make computations before we actually need them. 

This was requested as a feature by @tjhei.

The current design allows you to compute only what you ask for, and only the minimal amount of times, making sure that if the triangulation has changed between two consecutive calls, then the internal objects are actually updated.

Reply:
That’s a pity, I feels a bit hacky to me

Reply:
These are internal details. Of course you are free to come up (or suggest) a non-hacky solution that preserves the following features:

- no computation is done when the object is created the first time
- all internal objects are invalidated by changes to the underlying grid
- computation is only performed if necessary, i.e., if any of the following conditions are satisfied:
   - the underlying mesh has changed
   - the user requested something that was never computed before
   - the user explicitly asked to perform again a given computation (or all), for example because he or she knows that the data is invalid (say, e.g.,  because a vector stored in a `MappingQEulerian` has changed, and that mapping is used inside the `Cache`

The "hacky" thing you are complaining about satisfies the above criterion. Maybe we could remove the `const` qualifier to all getter functions (therefore the `mutable` to all internal objects). This however will require the `Cache` object to always be passed as a non-const object to all functions that will use it.

The design idea was to have `Cache` as a drop-in replacement to all functions that require read-only access to mesh-dependent arrays of objects (used vertices, positions, maps from vertices to cells, etc.). 

Most of these functions happen in `const` methods, and having a const qualifier to the getters of `Cache` seemed like a better idea.

Of course we can change that, if you feel this is the culprit...